src/devices/src/virtio/balloon/utils.rs

```diff
@@ -25,25 +26,33 @@ pub(crate) fn compact_page_frame_numbers(v: &mut [u32]) -> Vec<(u32, u32)> {
     let mut result = Vec::with_capacity(MAX_PAGE_COMPACT_BUFFER);
 
     // The most recent range of pages is [previous..previous + length).
-    let mut previous = v[0];
+    let mut previous = 0;
     let mut length = 1;
 
-    for page_frame_number in &v[1..] {
+    for pfn_index in 1..v.len() {
+        let page_frame_number = v[pfn_index];
+
+        // Skip duplicate pages.
+        if page_frame_number == v[pfn_index - 1] {
+            error!("Skipping duplicate PFN.");
+            continue;
+        }
+
         // Check if the current page frame number is adjacent to the most recent page range.
-        if *page_frame_number == previous + length {
+        if page_frame_number == v[previous] + length {
```

Comment:
This operation can overflow with the _right_ invalid input provided by a malfunctioning driver.

Reply:
This operation is what led me to make this change.

When we reviewed this algorithm, I thought about whether or not this operation needs to be a `checked_add`, but I decided it's not needed as a normal driver would only send distinct PFNs.

This makes the overflow impossible as for whatever value `v[previous]` has in the u32 range, we know there are at least `length` consecutive numbers greater than it in the array (the greatest so far being `page_frame_number`), since `v[previous]` is before all of them in the sorted array and `length` was incremented for each consecutive one.

The overflow can happen when the numbers are not distinct and we have multiple max values. 

Suppose you have an array of PFNs `[u32::MAX, u32::MAX]`. When running the old code, `previous` would be `u32::MAX`, length would be `1`. This will overflow as `previous` is **_never_** supposed to be `u32::MAX` unless the array has a length of 1.

This actually solves the overflow issue. I also added a unit test covering this.